trusted-neurons-alerts/governance.py

```diff
@@ -0,0 +1,79 @@
+import json
+import pathlib
+import tempfile
+
+import requests
+from ic import Canister
+from ic.agent import Agent
+from ic.candid import decode
+from ic.candid import Types
+from ic.certificate import lookup
+from ic.client import Client
+from ic.identity import Identity
+from ic.principal import Principal
+
+DFINITY_NEURON_ID = 27
+
+class GovernanceCanister:
+    """A simple client for querying the IC Mainnet Governance canister."""
+
+    def __init__(self):
+        """Create a new GovernanceCanister client."""
+        self.agent = Agent(Identity(), Client("https://ic0.app"))
+        self.principal = "rrkah-fqaaa-aaaaa-aaaaq-cai"
+        self.canister = None
+
+    def _version(self):
+        """Return the current git version of the Governance canister."""
+        paths = [
+            "canister".encode(),
+            Principal.from_str(self.principal).bytes,
+            "metadata".encode(),
+            "git_commit_id".encode(),
+        ]
+        tree = self.agent.read_state_raw(self.principal, [paths])
+        response = lookup(paths, tree)
+        version = response.decode("utf-8").rstrip("\n")
+        return version
+    
+    def _get_recent_ballots(self):
+        """Return latest 100 proposals voted by Neuron 27."""
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            version = self._version()
+            governance_did = pathlib.Path(tmpdirname) / "governance.did"
+            contents = requests.get(
+                f"https://raw.githubusercontent.com/dfinity/ic/{version}/rs/nns/governance/canister/governance.did",
+                timeout=10,
+            ).text
+            with open(governance_did, "w", encoding="utf8") as f:
+                f.write(contents)
+
+            governance_canister = Canister(agent=self.agent, canister_id=self.principal, candid=open(governance_did, encoding="utf8").read())
```

Comment:
Not sure if this is supported in `ic-py` but I know there is a way to fetch candid directly from a canister via metadata. It would remove the dependency to github and allow you to always be sure to have a correct version of candid

Reply:
This would be cool! I will look for it